### PR TITLE
[processor/metricsgeneration] Implement attribute matching functionality

### LINKED ISSUE #35425: [processor/metricsgeneration] Add option to match metric attributes for calculation operations
### Component(s)

processor/metricsgeneration

### Describe the issue you're reporting

**Context**
Currently the metrics generation processor simply grabs the first datapoint's value from the metric whose name matches the supplied `metric2` configuration option. This value is then used for generating all new metrics and datapoints.

**Current implementation's limitation**
The limitation here is that attributes are ignored when generating a new metric. An example of when this yields unexpected results is for the host metrics receiver's filesystem scraper. Let's say a user wants to calculate total size of the filesystem in bytes using the two metrics `system.filesystem.usage` and `system.filesystem.utilization`. This would be done by the following configuration:

```
experimental_metricsgeneration:
  rules:
    - name: system.filesystem.capacity
       unit: bytes
       type: calculate
       metric1: system.filesystem.usage
       metric2: system.filesystem.utilization
       operation: divide
```

The problem here is that the attributes of these two metrics may mismatch. One attribute is the filesystem's `mountpoint`. The resulting metric would incorrectly use the first datapoint's value, regardless of mountpoint. This would result in meaningless generated metrics.

**Proposal**
I propose adding a new configuration option, `match_attributes`, a `bool` that's disabled by default. This would preserve current functionality be default.

If this option is enabled, the two datapoints being compared must have identical attributes to create a resulting datapoint in the desired metric. If they're not an identical match, no action will be taken.

Resulting configuration:
```
experimental_metricsgeneration:
  rules:
    - name: system.filesystem.capacity
       unit: bytes
       type: calculate
       metric1: system.filesystem.usage
       metric2: system.filesystem.utilization
       match_attributes: true
       operation: divide
```

### PR DESCRIPTION
**Pull request by crobert-1**
#### Description
The metrics generation processor's default behavior is to get the first data point's value in the configured `metric2`, and use it for all calculations done with the given rule. As described in the issue, this may yield unexpected or incorrect results.

This introduces the feature gate `metricsgeneration.MatchAttributes`. When enabled, calculations will **only** be done on data points whose overlapping attributes match. This means that if both data points have a given attribute, their values must match.

#### Link to tracking issue
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35425

#### Testing
Added golden tests to ensure expected metrics are generated. Existing tests remain unchanged, showing this functionality did not impact other functions.

### CODE REVIEW

File: processor/metricsgenerationprocessor/processor_test.go [hunk 1]
```diff
@@ -392,130 +393,173 @@ func getOutputForIntGaugeTest() pmetric.Metrics {
...
-		// golden.WriteMetrics(t, filepath.Join("testdata", testCase.testDir, fmt.Sprintf("%s_%s", testCase.name, "expected.yaml")), got[0])
-		expected, err := golden.ReadMetrics(filepath.Join("testdata", testCase.testDir, fmt.Sprintf("%s_%s", testCase.name, "expected.yaml")))
-		assert.NoError(t, err)
-		assert.Len(t, got, 1)
-		err = pmetrictest.CompareMetrics(expected, got[0],
-			pmetrictest.IgnoreMetricDataPointsOrder(),
-			pmetrictest.IgnoreStartTimestamp(),
-			pmetrictest.IgnoreTimestamp())
-		assert.NoError(t, err)
+		t.Run(testCase.name, func(t *testing.T) {
```
Comment (crobert-1) [inline processor/metricsgenerationprocessor/processor_test.go:520]: I modified this test's setup so each test case can be run independently locally, and the test name in failures will be more obvious. The actual test functionality here remains largely unchanged, other than setting the feature gate value.

File: processor/metricsgenerationprocessor/utils.go [hunk 2]
```diff
@@ -50,36 +51,140 @@ func getMetricValue(metric pmetric.Metric) float64 {
...
+		}
+		return true
+	})
+
+	return attributesMatch
+}
+
+// generateScalarMetrics creates a new metric based on a scalar type rule and adds it to the scope metric.
+// The value for newly calculated metrics is always a floating point number.
+func generateScalarMetrics(rm pmetric.ResourceMetrics, operand2 float64, rule internalRule, logger *zap.Logger) {
```
Comment (crobert-1) [inline processor/metricsgenerationprocessor/utils.go:172]: This function is the original `generateMetrics` function, renamed to make the differing behavior obvious.

File: processor/metricsgenerationprocessor/utils.go [hunk 3]
```diff
@@ -126,15 +231,25 @@ func generateMetric(from pmetric.Metric, operand2 float64, operation string, log
 	return to
 }
 
-// Append the scope metrics with the new metric
-func appendMetric(ilm pmetric.ScopeMetrics, newMetric pmetric.Metric, name, unit string) pmetric.Metric {
-	metric := ilm.Metrics().AppendEmpty()
-	newMetric.MoveTo(metric)
+// Append the new metric to the scope metrics. This will only append the new metric if it
+// has data points.
+func appendNewMetric(ilm pmetric.ScopeMetrics, newMetric pmetric.Metric, name, unit string) {
```
Comment (crobert-1) [inline processor/metricsgenerationprocessor/utils.go:236]: The logic of only adding new metrics if they have valid data points was going to be copied in multiple places, so I decided to move it here to a centralized location.

File: processor/metricsgenerationprocessor/testdata/match_attributes/metrics_input.yaml
Comment (crobert-1) [inline processor/metricsgenerationprocessor/testdata/match_attributes/metrics_input.yaml:1]: I've attempted to make attribute values and data point values obviously different, to help scan the expected results for correctness. The important part is to make sure all original attributes are copied over (merging the attributes of metric 1 and metric 2), and to ensure the "overlapping attributes" functionality is handled properly.

File: processor/metricsgenerationprocessor/utils.go [hunk 4]
```diff
@@ -50,36 +51,140 @@ func getMetricValue(metric pmetric.Metric) float64 {
...
+						case pcommon.ValueTypeDouble:
+							newDP.Attributes().PutDouble(k, v.Double())
+
+						case pcommon.ValueTypeInt:
+							newDP.Attributes().PutInt(k, v.Int())
+						case pcommon.ValueTypeBool:
+							newDP.Attributes().PutBool(k, v.Bool())
+						default:
+							newDP.Attributes().PutStr(k, v.Str())
+						}
```
Comment (dmitryax) [inline processor/metricsgenerationprocessor/utils.go:133]: No need to track all possible types
```suggestion
						v.CopyTo(newDP.Attributes().PutEmpty(k))
```

Comment (crobert-1) [reply processor/metricsgenerationprocessor/utils.go:133]: Thanks Dmitrii, that's much better 👍

**Comment by atoulme** [comment]
I just rebased off main to fix go.mod/go.sum conflicts.

**Review by dmitryax — APPROVED** [review]

**Pull request merged**